src/AI/ModularAI.java

```diff
@@ -0,0 +1,181 @@
+  // List of possible AI modes, in order of precedence
+  protected ArrayList<AIModule> aiPhases;
+  // Used to determine what the default priority for unit actions is
+  protected Comparator<Unit> unitOrderSetter = new AIUtils.UnitCostComparator(false);
+
+  private StringBuffer logger = new StringBuffer();
+  protected int turnNum = 0;
+
+  public ModularAI(Commander co)
+  {
+    myCo = co;
+  }
+
+  @Override
+  public void initTurn(GameMap gameMap)
+  {
+    logger = new StringBuffer(); // Reset at the start of the turn so the AI's action log stays in memory between turns for review
+    ++turnNum;
+    // Create a list of every property we don't own, but want to.
+    unownedProperties = new ArrayList<XYCoord>();
+    for( int x = 0; x < gameMap.mapWidth; ++x )
+    {
+      for( int y = 0; y < gameMap.mapHeight; ++y )
+      {
+        XYCoord loc = new XYCoord(x, y);
+        if( gameMap.getLocation(loc).isCaptureable() && myCo.isEnemy(gameMap.getLocation(loc).getOwner()) )
+        {
+          unownedProperties.add(loc);
+        }
+      }
+    }
+
+    for( AIModule phase : aiPhases )
+    {
+      phase.initTurn(gameMap);
+    }
+  }
+
+  protected void log(String message)
+  {
+    System.out.println(message);
+    logger.append(message).append('\n');
+  }
+
+  @Override
+  public GameAction getNextAction(GameMap gameMap)
+  {
+    ArrayList<Unit> eligibleUnits = new ArrayList<Unit>();
+    for( Unit unit : myCo.units )
+    {
+      if( unit.isTurnOver || !gameMap.isLocationValid(unit.x, unit.y) )
+        continue; // No actions for units that are stale or out of bounds.
+      eligibleUnits.add(unit);
+    }
+
+    GameAction nextAction = null;
+    PriorityQueue<Unit> unitQueue = new PriorityQueue<Unit>(13, unitOrderSetter);
+
+    // Each module gets a shot at the whole unit queue, in order.
+    // If no action's found, we're done
+    for( AIModule phase : aiPhases )
+    {
+      unitQueue.clear();
+      unitQueue.addAll(eligibleUnits);
+
+      nextAction = phase.getNextAction(unitQueue, gameMap);
+      if( null != nextAction )
+        break;
+    }
+    log(String.format("Action: %s", nextAction));
+    return nextAction;
+  }
+
+  public static interface AIModule extends Serializable
+  {
+    public GameAction getNextAction(PriorityQueue<Unit> unitQueue, GameMap map);
+    public default void initTurn(GameMap gameMap) {}
+  }
+
+  public static class PowerActivator implements AIModule
+  {
+    private static final long serialVersionUID = 1L;
+    public Commander myCo;
+    public final int aiPhase;
+
+    public PowerActivator(Commander co, int phase)
+    {
+      myCo = co;
+      aiPhase = phase;
+    }
+
+    public boolean checked = false;
+    @Override
+    public void initTurn(GameMap gameMap) { checked = false; }
+    @Override
+    public GameAction getNextAction(PriorityQueue<Unit> unitQueue, GameMap map)
+    {
+      if( checked )
+        return null;
+      checked = true;
```

Comment:
Won't this make it so this module can only ever activate an ability at the start of a turn?

Reply:
It means you only ever check for the ability once.
If you place it at the end of the list of modules, you're guaranteeing that if you *reach* the module, you're already otherwise going to end your turn anyway.
I only foresee problems with this if we introduce the ability to use multiple CO powers in one turn.